README.md
[![ci-badge][]][ci] [![license-badge][]][license] [![docs-badge][]][docs] [![rust badge]][rust link]

# urbandictionary.rs

Unofficial Rust crate for the Urbandictionary API.

[Documentation][docs]

### Installation

This library requires at least Rust 1.21.

Add the following dependency to your Cargo.toml:

```toml
urbandictionary = "0.3"
```

And include it in your project:

```rust
extern crate urbandictionary;
```

### Examples

Using `hyper` with the `hyper-tls` HTTPS connector, retrieve a list of
definitions for a word and print the example of the second definition if it
exists:

```rust
extern crate futures;
extern crate hyper;
extern crate hyper_tls;
extern crate tokio_core;
extern crate urbandictionary;

use futures::Future;
use hyper::client::{Client, HttpConnector};
use hyper_tls::HttpsConnector;
use std::error::Error;
use tokio_core::reactor::Core;
use urbandictionary::HyperUrbanDictionaryRequester;

fn try_main() -> Result<(), Box<Error>> {
    let mut core = Core::new()?;
    let client = Client::configure()
        .connector(HttpsConnector::new(4, &core.handle())?)
        .build(&core.handle());

    let done = client.definitions("cat").and_then(|response| {
        if let Some(definition) = response.definitions.get(1) {
            println!("Examples: {}", definition.example);
        }

        Ok(())
    }).map_err(|_| ());

    core.run(done).expect("Error running core");

    Ok(())
}

fn main() {
    try_main().unwrap();
}
```

Using reqwest, print the definition of the word `"cat"`:

```rust
extern crate reqwest;
extern crate urbandictionary;

use reqwest::blocking::Client;
use std::error::Error;
use urbandictionary::ReqwestUrbanDictionaryRequester;

fn try_main() -> Result<(), Box<Error>> {
    let client = Client::new();
    let response = client.define("cat")?;

    if let Some(definition) = response {
        println!("The definition of cat is: {}", definition.definition);
    } else {
        println!("No definition found");
    }

    Ok(())
}

fn main() {
    try_main().unwrap();
}
```

### License

License info in [LICENSE.md]. Long story short, ISC.

[ci]: https://travis-ci.org/zeyla/urbandictionary.rs
[ci-badge]: https://img.shields.io/travis/zeyla/urbandictionary.rs.svg?style=flat-square
[docs]: https://docs.rs/crate/urbandictionary
[docs-badge]: https://img.shields.io/badge/docs-online-2020ff.svg?style=flat-square
[license.md]: https://github.com/zeyla/urbandictionary.rs/blob/master/LICENSE.md
[license]: https://opensource.org/licenses/ISC
[license-badge]: https://img.shields.io/badge/license-ISC-blue.svg?style=flat-square
[rust badge]: https://img.shields.io/badge/rust-1.21+-93450a.svg?style=flat-square
[rust link]: https://blog.rust-lang.org/2017/10/12/Rust-1.21.html


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [0.3.1] - 2020-01-30

This updates Reqwest to version 0.10

### Changed

-   Update Reqwest to version 0.10 while still using the blocking api

## [0.3.1] - 2018-07-28

This fixes a bug causing deserializations to fail.

### Fixed

-   Make `Response::tags` optional ([@drklee3]) [c:e781593]

[c:e781593]: https://github.com/zeyla/urbandictionary.rs/commit/e781593b2ecade8552c0320aeae3fa4550614be7

## [0.3.0] - 2017-01-16

### Changed

-   Rewrite the library; now includes trait requester implementations for
    hyper v0.11's Client and Reqwest v0.8's client

## [0.2.0] - 2017-08-02

### Changed

-   The `define` and `definitions` functions now accepted a string slice;
-   A new struct has been made which is a thin wrapper over a Hyper Client;
-   Now uses serde_derive;
-   Dependencies updated.

## [0.1.2] - 2017-01-15

### Added

### Changed

-   `error` is no longer a public module; use the re-exports instead [BC break];
-   models are now deserialized via serde_codegen;
-   documentation is now enforced on all items.

## [0.1.1] - 2016-08-21

### Added

### Changed

-   Make every field public

## [0.1.0] - 2016-08-21

Initial commit.

[@drklee3]: https://github.com/drklee3
[0.3.1]: https://github.com/zeyla/urbandictionary.rs/compare/v0.3.0...v0.3.1
[0.3.0]: https://github.com/zeyla/urbandictionary.rs/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/zeyla/urbandictionary.rs/compare/v0.1.2...v0.2.0
[0.1.2]: https://github.com/zeyla/urbandictionary.rs/compare/v0.1.1...v0.1.2
[0.1.1]: https://github.com/zeyla/urbandictionary.rs/compare/v0.1.0...v0.1.1


src/lib.rs
//! [![ci-badge][]][ci] [![license-badge][]][license] [![docs-badge][]][docs] [![rust badge]][rust link]
//!
//! # urbandictionary.rs
//!
//! Unofficial Rust crate for the Urbandictionary API.
//!
//! [Documentation][docs]
//!
//! ### Installation
//!
//! Add the following dependency to your Cargo.toml:
//!
//! ```toml
//! urbandictionary = "0.3"
//! ```
//!
//! And include it in your project:
//!
//! ```rust,no_run
//! extern crate urbandictionary;
//! ```
//!
//! ### Examples
//!
//! Using `hyper` with the `hyper-tls` HTTPS connector, retrieve a list of
//! definitions for a word and print the example of the second definition if it
//! exists:
//!
//! ```rust,no_run
//! # #[cfg(feature = "hyper-support")]
//! extern crate futures;
//! # #[cfg(feature = "hyper-support")]
//! extern crate hyper;
//! # #[cfg(feature = "hyper-support")]
//! extern crate hyper_tls;
//! # #[cfg(feature = "hyper-support")]
//! extern crate tokio_core;
//! extern crate urbandictionary;
//!
//! # use std::error::Error;
//! #
//! # #[cfg(feature = "hyper-support")]
//! # fn try_main() -> Result<(), Box<Error>> {
//! #
//! use futures::Future;
//! use hyper::client::{Client, HttpConnector};
//! use hyper_tls::HttpsConnector;
//! use tokio_core::reactor::Core;
//! use urbandictionary::HyperUrbanDictionaryRequester;
//!
//! let mut core = Core::new()?;
//! let client = Client::configure()
//!     .connector(HttpsConnector::new(4, &core.handle())?)
//!     .build(&core.handle());
//!
//! let done = client.definitions("cat").and_then(|response| {
//!     if let Some(definition) = response.definitions.get(1) {
//!         println!("Examples: {}", definition.example);
//!     }
//!
//!     Ok(())
//! }).map_err(|_| ());
//!
//! core.run(done).expect("Error running core");
//! #     Ok(())
//! # }
//! #
//! # fn main() {
//! #    #[cfg(feature = "hyper-support")]
//! #    try_main().unwrap();
//! # }
//! ```
//!
//! Using reqwest, print the definition of the word `"cat"`:
//!
//! ```rust,no_run
//! # #[cfg(feature = "reqwest-support")]
//! #
//! extern crate reqwest;
//! extern crate urbandictionary;
//!
//! # use std::error::Error;
//! #
//! # #[cfg(feature = "reqwest-support")]
//! # fn try_main() -> Result<(), Box<Error>> {
//! #
//!
//! use reqwest::Client;
//! use urbandictionary::ReqwestUrbanDictionaryRequester;
//!
//! let client = Client::new();
//! let response = client.define("cat")?;
//!
//! if let Some(definition) = response {
//!     println!("The definition of cat is: {}", definition.definition);
//! } else {
//!     println!("No definition found");
//! }
//! #     Ok(())
//! # }
//! #
//! # fn main() {
//! #     #[cfg(feature = "reqwest-support")]
//! #     try_main().unwrap();
//! # }
//! ```
//!
//! ### License
//!
//! License info in [LICENSE.md]. Long story short, ISC.
//!
//! [ci]: https://travis-ci.org/zeyla/urbandictionary.rs
//! [ci-badge]: https://img.shields.io/travis/zeyla/urbandictionary.rs.svg?style=flat-square
//! [docs]: https://docs.rs/crate/urbandictionary
//! [docs-badge]: https://img.shields.io/badge/docs-online-2020ff.svg?style=flat-square
//! [LICENSE.md]: https://github.com/zeyla/urbandictionary.rs/blob/master/LICENSE.md
//! [license]: https://opensource.org/licenses/ISC
//! [license-badge]: https://img.shields.io/badge/license-ISC-blue.svg?style=flat-square
//! [rust badge]: https://img.shields.io/badge/rust-1.21+-93450a.svg?style=flat-square
//! [rust link]: https://blog.rust-lang.org/2017/10/12/Rust-1.21.html
#![deny(missing_docs)]

#[macro_use]
extern crate serde_derive;

extern crate serde;
extern crate serde_json;

#[cfg(feature = "futures")]
extern crate futures;
#[cfg(feature = "hyper")]
extern crate hyper;
#[cfg(feature = "hyper-tls")]
extern crate hyper_tls;
#[cfg(feature = "reqwest")]
extern crate reqwest;

pub mod bridge;
pub mod model;

mod error;

pub use crate::error::{Error, Result};

#[cfg(feature = "hyper-support")]
pub use crate::bridge::hyper::UrbanDictionaryRequester as HyperUrbanDictionaryRequester;
#[cfg(feature = "reqwest-support")]
pub use crate::bridge::reqwest::UrbanDictionaryRequester as ReqwestUrbanDictionaryRequester;
